### ISSUE TITLE
Run-time format selection

### ISSUE THREAD

**Issue by zah** [status-im/nim-chronicles#75]
Currently, Chronicles allows you to specify the output format associated with a particular sink at compile-time, but some apps may require the format to be specified at run-time on the command-line or within a config file.

Adding support for this boils down to the creation of a new `DynamicLogRecord` type with a generic tuple parameter `FormatsList`. At compile-time, the user will specify the list of formats that will be compiled in the program. At run-time, each chronicles stream using dynamic log records as its output type will also carry a numeric index of the currently active format. Each written property in a log statement uses the active index to perform a lookup in a table of functions associated with the logged property type and thus the right output type is selected at run-time (similar to a virtual function call in C++).

**Comment by disruptek** [comment]
Why not use strings instead of numeric indexes?

**Comment by zah** [comment]
The numeric indices will be more efficient. But perhaps you are not imagining this correctly. It's supposed to work very similarly to a C++ vtable. Would you propose these to be backed by string hash tables?

Of course the configuration, and some parts of the APIs will rely on strings to identify the different formats.

**Comment by disruptek** [comment]
Yes.  I think a string is a much more useful type to pass around.  What happens under the covers isn't important to me as a user.

**Comment by disruptek** [comment]
To put this another way, I don't see the value in having a seemingly random index in my dynamic log record output.  The means by which I make sense of this index elsewhere will be fairly brittle, right?

**Comment by zah** [comment]
The `index` is purely an implementation mechanism. It won't appear in the log output. It appears only in the memory of the program as a field of a particular log stream. Its only purpose is to select the log formatting function that will transform the strongly-typed logged properties to the serialized output of the particular formats offered as a run-time choice (e.g. json, xml, textlines, etc).

The high-level API can work with symbol names (because each format in nim-serialization has an associated symbol).

```nim
myLogStream.setOutputFormat(Json)
```

**Comment by disruptek** [comment]
Ah, gotcha; sounds good!

**Comment by Benjamin-Lee** [comment]
This would be a great addition. For my use case, I'm building a CLI tool that will need the ability to configure the format at run time since I'm planning on distributing it in a way that doesn't require the users (bioinformaticians) manually compile.

**Comment by arnetheduck** [comment]
`setLogEnabled` can now be set [per sink](https://github.com/status-im/nim-chronicles/pull/172) - this allows disabling sinks individually, which works as a format selection facility (at compile time, create sinks for each format then enable the ones you want)
